model/canvas.cc

```diff
@@ -148,7 +148,7 @@ namespace minsky
           if (auto to=closestInPort(x,y)) {
             model->addWire(static_cast<shared_ptr<Port>&>(fromPort),to);
             fromPort.reset();
-            if(!to->item().inputDefined()) {
+            if(to->item().inputIsScalar()) {
```

Comment:
Help me understand this... As I understand our options are scalar (0 dimensional, single value) and tensor (multi-dimensional range/field/cube of values). We're checking this prior to a minsky().reset() call to see if initialization is required... if the connected object's input was already tensor prior to a reset, why is initialization required?
Do I understand correctly that the "default" input value (that is set prior to connecting anything) is a scalar of zero? There is no "null" option in this value type?

Reply:
I don't seem to get consistent results no matter what I do here... the first time I connect a Ravel's input to a CSV import, it has vv == null, then when I delete the Ravel and create a new one and connect it, the Rank of the second port (first input) is 3...

Reply:
Technically, a scalar is also a rank0 tensor. The other options are that the input port is not connected, or that it is connected, but has no variableValue associated with it. What we wanted to do was to only call Minsky::reset when the input being connected was known to be rank 1 or higher tensor. But ultimately, we decided not to even make that distinction.

Reply:
I think one thing that makes me confused might be that information on the rank of input is transmitted between objects already before minsky().reset() takes place, am I guessing that correctly? Previously I have been assuming that there is one datatype that is transmitted post-calculation with both the range of values and their rank, but if these are transmitted separately I have less trouble understanding what you're saying.
What I then still do not understand is why the first time I connect a Ravel to a CSV input object, the variableValue on the input port comes in as null... delete the Ravel, create a new one and connect it and it will be a rank 3 object... why this inconsistency? Isn't rank 3 input what we're consistently expecting to see here, if the rank information is transmitted immediately from the input object?